Gaussian naive bayes

rlouf opened the pull request:
I implemented the Gaussian Naive Bayes algorithm. I added tests, a notebook and updated the docs.

Commit:
Implement the Gaussian Naive Bayes model

Commit:
Add tests for the Gaussian Naive Bayes model

Commit:
Add notebook to describe the GaussianNaiveBayes' functionalities

Commit:
Style changes in code and doctrings

Part of the code didn't follow the PEP8 conventions very closely, there
were imports that were not used, and the docstrings did not use the
conventions that had already been adopted in the library.

rlouf commented:
Closes #4

Commit:
Update CHANGELOG and AUTHORS.txt

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,141 @@
+from pymc3 import summary
+from sklearn.model_selection import train_test_split
+
+from pymc3_models.exc import  PyMC3ModelsError
```
Extra space after `import`.

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,141 @@
+
+import numpy as np
+import scipy.stats
+from pymc3 import summary
```
Are you using this import?

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,141 @@
+
+from pymc3_models.exc import  PyMC3ModelsError
+from pymc3_models import GaussianNaiveBayes
+
```
Extra blank line between the imports and the class per PEP8 (https://www.python.org/dev/peps/pep-0008/#imports).

rlouf replied:
Are you refering to the "2 blank lines before classes" rule?

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class GaussianNaiveBayesPredictProbaTest(GaussianNaiveBayesTestCase):
+    """ Test the predict_proba() method.
```
Since your testcases are very explicitly named (yea!), you don't need docstrings.

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,141 @@
+        strategies to test probabilistic code.
+        """
+
+        print('') # Berk
```
Could you make this comment more explicit? I'm not sure what you mean by `Berk`.

rlouf replied:
Same as above. I remove it.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+        Model
+        -----
+
+        We note $x_{jc}$ the value of the j-th element of the data vector $x$
```
Were you able to get the latex to render locally? My cursory research suggests we may need to put some prefixes to get Sphinx to pick up on it (http://www.sphinx-doc.org/en/stable/ext/math.html).

rlouf replied:
You were right, I changed the prefix and Latex equations now render ok. The only issue I have with it is that it's very ugly; I'll try to see if we can change that by tweaking the settings.

parsing-science reviewed (changes requested):
@rlouf: This looks like it's going to be really cool! Thanks for being the first to add a model. I did a quick check through your code and will do a more thorough review soon.

Just as a heads up, I tend to care a lot about formatting and PEP8 so a lot of my comments are in that vein.

Two main things:
1) There were some inconsistencies in your coding style that I think could be smoothed over. I left comments to point out examples.
2) I found some PEP8 errors in the code you submitted. I pointed out a few, but it would be helpful if you could run the code through a linter before the next review.

Thanks again!

rlouf commented:
Thanks! What linter would you recommend? I've used `pyflakes` and `pylint` but they missed most of the things you pointed out.

rlouf commented on CHANGELOG.md:
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
+## [1.1.1] - 2018-02-17
```
That's correct.

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+import functools as ft
+import numpy as np
+import scipy.stats
+import theano
```
Yes, that's more readable indeed.

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+
+
+
+# Gaussian Naive Bayes
```
As you can see, I use these comments as dividers in the code (there is a "Functions" section at the bottom). I was planning on implementing Binary and Multinomial Naive Bayes in the same file, so left this comment in the source. I can remove it if you think that's a problem.

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+
+        Parameters
+        ----------
+        X (np.array): Numpy array of shape [n_samples, n_features] containing the 
```
True, I didn't double-check. Having said that, I prefer the following style where name & type are on the same line and explanation on the second line: https://github.com/networkx/networkx/blob/master/networkx/algorithms/shortest_paths/astar.py
I find it less clutered. I'll follow the current conventions for now, and we can alway discuss about docstring style later.

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+        The current instance of the model.
+
+        """
+        self.n_samples, self.n_features = X.shape 
```
I was trying to follow Scikit-learn's naming conventions (see the code for [k-means](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/k_means_.py) for instance). I personally find they make sense. I'm aware it's a subjective notion, but I think it would be wise to adopt the same conventions as sk-learn here.

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+            if len(predictions) == 0:
+                predictions = prob_per_class
+            else:
+                predictions = np.vstack((predictions, prob_per_class)) # Mutability, berk.
```
I wasn't happy with the way this was coded when I first wrote the code, and this comment was only intended to be seen by me. I took it off.

rlouf commented:
Also, not directly related to this: most python packages adopt the `snake_case` naming convention for files. Wouldn't it make sense to do the same here?

rlouf commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,269 @@
+# -----------------------------------------------------------------------------
+
+class GaussianNaiveBayes(BayesianModel):
+    """ Naive Bayes classification built using PyMC3.
```
Wouldn't it make sense to follow [PEP8's recommendations](https://www.python.org/dev/peps/pep-0008/#module-level-dunder-names) here too?

Commit:
Add LateX support in documentation

rlouf commented:
I implemented most of the requested changes. I am not sure the docstrings in the rest of the package follow PEP8. I left mine as they were; I can follow the rest of the package, and we may address this in an another PR.

parsing-science commented:
First of all, I appreciate your enthusiasm for this project. I'm glad you have a lot of ideas to contribute. 

However, I would ask that you please follow the contributor guidelines (https://github.com/parsing-science/pymc3_models/blob/master/CONTRIBUTING.rst), particularly the section about following existing conventions. I made thoughtful deliberate decisions about style and the code itself so I'd appreciate if you respected those. 

If you want to refactor your code to match the rest of the repo, I'll be happy to take another look at your PR.

Commit:
Update the Jupyter Notebook

I edited the content of the notebook, adding a graphical representation
of the Gaussian Naive Bayes model and re-run the entire notebook to take
the changes into account.

Commit:
Amend style in code and docstrings

rlouf commented:
Sorry, I wasn't questioning your choices. Seeing the references to scikit-learn and numpy in the README and CONTRIBUTING, I started writing code with those in ming. I realised after the fact that the conventions adopted here were slightly different. Not a problem for me, I was happy to refactor to match the rest of the repo.

I ran pyflakes and pycodestyle on both the model and tests. All I got were unused variables in the declaration of the model, and lines longer than 79 characters---they are all shorter than 100 characters thought.

twiecki commented:
This looks great! @parsing-science are there any outstanding issues you see?

parsing-science commented:
I've been busy with some personal stuff in the last few months. I'll try to take a look in the next few weeks.

rlouf commented:
Is there anything I can do to help? I started implementing other models, but I would like to have this PR merged before making them PR-ready.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+__all__ = ["GaussianNaiveBayes"]
+
+
+# Gaussian Naive Bayes
```
I don't think you need this comment since it's the same as the classname.

rlouf replied:
Fixed.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+from pymc3_models.models import BayesianModel
+
+
+__all__ = ["GaussianNaiveBayes"]
```
Since we're importing the models in the __init__ file, I don't think we need to keep this.

rlouf replied:
It is not necessary, but I think it improves readability when there are several classes in the same file. I can remove it for now, and put it back If we ever add another class (MultinomialNaiveBayes, etc.) ?

parsing-science replied:
sound good!

rlouf replied:
Done.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+        """
+        Creates and returns the PyMC3 model.
+
+        We note :math:`x_{jc}` the value of the j-th element of the data vector :math:`x`
```
Did you get the math to render? I don't see it when I look on github.

rlouf replied:
The math formatting is now correct for Sphinx. I got to render it, but I had to change some parameters in the Sphinx config. I’ll adress this in another PR.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+        Creates and returns the PyMC3 model.
+
+        We note :math:`x_{jc}` the value of the j-th element of the data vector :math:`x`
+        conditionned on x belonging to the class :math:`c`. The Gaussian Naive Bayes
```
Typo: `conditioned`

rlouf replied:
Thanks! There must be a bunch of typos left, I'll proof-read before re-submitting.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+            # Assign classes to data points
+            z = pm.Categorical('z', pi, shape=self.num_training_samples, observed=y)
+
+            # The components are independant and normally distributed
```
Typo: `independent`

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+            # Priors
+            alpha = np.ones(self.num_cats)
+            pi = pm.Dirichlet('pi', alpha, shape=self.num_cats)
+            mu = pm.Normal('mu', 0, 100, shape=(self.num_cats, self.num_pred))
```
Can you explicitly call mu=0 and sd=100 in these distributions? I got bitten when the PyMC3 API switched from tau to sd as default so I prefer to be explicit.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+        y : numpy array, shape [num_training_samples,]. Contains the category of the data points.
+
+        inference_type : string, specifies which inference method to call.
+            Default is 'advi'. Currently, only 'advi' and 'nuts'are implemented.
```
Need space between `'nuts'` and `are`

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+
+        Returns
+        -------
+        The current instance of the PyMC3 model.
```
This is potentially confusing because it returns the class instance, not the PyMC3 model instance. Could you clarify this to make it clear it's an instance of GaussianNaiveBayes and not pm.Model?

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+        predictions = np.array([])
+        for x in X:
+            prob_per_sample = scipy.stats.norm(self.trace['mu'], self.trace['sigma']).pdf(x)
+            prob_per_feature = [np.sum(prob_per_sample[:, :, i], axis=0)/len(self.trace['mu'])
```
Since this line is too long, could you format it like this:
```
prob_per_feature =  [
      np.sum(prob_per_sample[:, :, i], axis=0)/len(self.trace['mu']) for i in range(self.num_pred)
]
```

I find that easier to read.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+                                for i in range(self.num_pred)]
+            prob_per_class = normalize(ft.reduce(lambda x, y: x*y, prob_per_feature))
+            if len(predictions) == 0:
+                predictions = prob_per_class
```
it's a bit confusing that predict_proba returns something called `predictions` and predict returns `pred`. I would expect predict to return `predictions`.  Any ideas for a clearer name for the variable here?

rlouf replied:
posterior_prediction?

parsing-science replied:
sure!

rlouf replied:
Technically, `predict` returns a point estimation of the full posterior distribution and `predict_proba` returns the Bayesian posterior prediction. I get how the current names make sense to people who come from traditional ML, but this may bite us further down the line if we ever introduce loss functions for point estimations.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+
+        Returns
+        -------
+        A float representing the score of the predictions.
```
Maybe put `representing the accuracy score` to be explicit about what scoring method you use.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,283 @@
+        self.num_training_samples = params['num_training_samples']
+
+
+# Functions
```
Let's move this to another file called something like pymc_models/common/utils.py. That way it will be available to any other models that might need this.

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,134 @@
+        shutil.rmtree(self.test_dir)
+
+
+# Test fit
```
I don't think this comment is necessary since the method is called `test_fit`.

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,134 @@
+        """
+        Test the model initialization and fit
+
+        Currently, only the sign of infered parameters is checked
```
Typo: `inferred`

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,134 @@
+        self.assertEqual(stds.shape, self.Y_test.shape)
+
+    def test_predict_proba_raises_error_if_not_fit(self):
+        """
```
I think the docstring is unnecessary since the model name is explicit.

parsing-science reviewed (changes requested):
@rlouf: Finally got a chance to get some comments back to you! 

Things are looking really good. I've made some bug fixes so please merge master back into your branch when you get a chance.

Your tests don't pass anymore, and I think you're running into the same issue that I had recently (https://github.com/pymc-devs/pymc3/issues/2985). You probably need to specify the dtype of your `y` in the PyMC3 model.

rlouf commented:
I'm still waiting for the tests to finish, please wait before reviewing the changes.

Commit:
Make all tests pass

Tests were failing for several reasons, mostly typos and mistakes in the
name of variables. Tests now all pass.

Commit:
Move normalize function to utils

rlouf commented:
All right, I made all the requested changes and the tests now pass (your fix did the trick). I also added a picture of the graphical model corresponding to Naive Bayes for the sake of being fancy. If this looks ok to you I can do a final rebase to clean this up before merging.

parsing-science commented on pymc3_models/models/NaiveBayes.py:
```diff
@@ -0,0 +1,273 @@
+        for x in X:
+            prob_per_sample = scipy.stats.norm(self.trace['mu'], self.trace['sigma']).pdf(x)
+            prob_per_feature = [
+                    np.sum(prob_per_sample[:, :, i], axis=0)/len(self.trace['mu'])
```
Could you format this as I suggested? My linter is complaining when it's on two lines.

rlouf replied:
Yes sorry, I totally missed that comment :man_facepalming:

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,123 @@
+
+        # TODO: How do you write tests for a stochastic mode?
+        # TODO: Diagnose the sampling with a reasonable sampling size?
+        np.testing.assert_equal(
```
These long lines are inconsistent about where the closing paren is. 
I would prefer: 
```
np.testing.assert_equal(
            np.sign(self.pi),
            np.sign(self.test_GNB.trace['pi'].mean(axis=0))
)
```

rlouf replied:
:+1:

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,123 @@
+        with self.assertRaises(PyMC3ModelsError) as no_fit_error:
+            test_GNB = GaussianNaiveBayes()
+            test_GNB.predict_proba(self.X_train)
+        expected = 'Run fit() on the model before predict()'  # precarious, define a new Exception type
```
What do you mean by this `precarious` comment?

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,123 @@
+    def test_score_scores(self):
+        print('')
+        score = self.test_GNB.score(self.X_test, self.Y_test)
+        # What to test for?
```
This is not the most informative. If you're not sure, could you just put in something more like: ```TODO: Figure out how to test the score function```.

Additionally, If there's no test, I would just put use the skip decorator (https://docs.python.org/3/library/unittest.html#unittest.skip): ```@unittest.skip('Not implemented yet')```

There's no reason to run the scoring if there's no test since it just takes up time.

rlouf replied:
You're right!

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,123 @@
+
+
+class GaussianNaiveBayesSaveAndLoadTestCase(GaussianNaiveBayesTestCase):
+    def test_save_and_load_work_correctly(self):
```
It looks like this test doesn't test anything?

rlouf replied:
True, somehow I didn't write the part that actually tests something ><

parsing-science reviewed (changes requested):
This looks great! Just a few minor comments. I was able to build the docs locally, and I really like the way the math looks. Thanks for figuring that out. 

Once you merge in master and make these final changes, we should be good to merge.

Commit:
Fix formatting issue

Commit:
Fix issues with the tests

SaveAndLoad did not test anything, and TestScore was not implemented. I
also removed a bunch of unnecessary comments.

rlouf commented:
I (finally) addressed all the issues you raised, rebased my branch onto master and update the date in the changelog. This should be ready to merge if you approve of the changes.

parsing-science commented on tests/models/test_NaiveBayes.py:
```diff
@@ -0,0 +1,135 @@
+        # Generate data
+        ## Priors
+        self.alpha = np.ones(self.num_cats)
+        self.pi = np.random.dirichlet(self.alpha)
```
I just set up Travis CI to run the unittests and realized I should have been setting the random seed. Could you do that in this file? Like here, https://github.com/parsing-science/pymc3_models/blob/master/tests/models/test_LinearRegression.py#L23.

parsing-science reviewed (approved):
@rlouf: finally had time to get back to this library! Your PR looks good.

I've made a bunch of changes in master though including adding a linter and Travis CI. I think I can merge master into your fork if I pull it down and merge it or you could merge in master yourself. Do you have a preference?

rlouf commented:
Cool! I’ll rebase my branch against master, make the changes, clean the history a little bit and it should be ready to merge.

Commit:
Set seed for the tests

parsing-science commented:
@rlouf: thanks for doing this! It looks like the linting is failing so I stopped the unittests so you can fix that stuff up first.

rlouf commented:
@parsing-science Thanks! Makes sense, I was not using flake8 as a linter, good thing you enforced that with CI! Will make the necessary changes ASAP.

rlouf commented:
Done! Waiting for Travis to do its job :)

parsing-science commented:
Actually I pulled down your branch and ran some of the tests so I'm going to merge this. Thanks!

rlouf commented:
Thanks! This was a great learning experience, thank you for bearing with me. I hope to be able to contribute more models soon.
